README.md
# Cardiac Risk App #
The widely-used [Reynolds Risk Score](http://www.reynoldsriskscore.org/) is used to estimate the 10-year cardiovascular risk of an individual. For patients and clinicians alike, this calculation is often reported in an esoteric, hard-to-read lab report.

For our model of an improved Reynolds Risk Score reporting SMART App, we took an [inspired Creative Commons design](http://www.informationisbeautiful.net/2010/visualizing-bloodtests/), the work of designer [David McCandless](http://www.davidmccandless.com/), whose “Blood Work Cardiology Result” appeared in the Wired Magazine article ["The Blood Test Gets a Makeover"](http://www.wired.com/2010/11/ff_bloodwork/all) (Wired 18.12, December 2010) and was published under a Creative Commons license. The resulting SMART Cardiac Risk app presents relevant patient vitals and lab measurements and the calculated Reynolds Risk Score, along with a succinct, patient-friendly explanation for each result. The app’s presentation produces a highly attractive document to hand to a patient. Furthermore, the SMART Cardiac Risk app also offers simulation: the clinician (or patient) can make changes to one or more of their smoking and family hustory and blood pressure to see how the patient’s current Reynolds Risk Score could be improved.

App Design:
![Sample App Design Screen Shot](CardiacRiskApp.png)

# Running Locally #

In order to run the Cardiac Risk app locally, you need to run a local webserver.
You can use the [http-server](https://www.npmjs.com/package/http-server) from [NPM](https://www.npmjs.com/).
Additionaly, you will need [fhirclient](https://www.npmjs.com/package/fhirclient) to run the fhir-client dependency.

1. Install [NPM](https://github.com/npm/npm#super-easy-install)
2. You would need http-server (to run the project locally) and fhirclient. Run ```npm install``` and all the necessary node_modules will be installed on your system.
3. Skip this step if Step 2 was a success : Run ```npm install http-server -g``` for installing http-server and ```npm install fhirclient``` for fhir-client.
4. Clone down the project and update location for your node_modules/fhirclient/fhir-client.min.js file in index.html and launch.html.
5. Navigate to the directory containing ```launch.html```
6. Run ```http-server -c-1 -p 8000```

# Build #

This project uses [Grunt](http://gruntjs.com/) to build the application. This will lint the project and create the build file.

1. If you have ran the step 2 from run locally then you dont need to install grunt. ( Incase if grunt fails to run install Grunt CLI: ```npm install -g grunt-cli```).
2. Just typing ```grunt``` will show all possible available tasks as specified in the Gruntfile.js
3. Run the following command while in the root directory of the application: ```grunt build```
4. This should create the cardiac_risk.css and cardiac_risk.js. It will also run minification jobs on these files and will lint them.

These files are referenced in the HTML file index.html.
The ```build/``` folder in the project acts as the final version's of these files to be used where as the ```src/``` folder is more for development.

# App Development/Improvements Process #
1. Make all the necessary changes to the files available in the ```src/``` directory.
2. Once changes are made run ```grunt build```
3. Run the application locally.

# Testing #

This project uses the [Mocha](http://mochajs.org/) testing framework to perform asynchronous unit testing.

To run the Mocha framework:

1. Follow step 2 from Running Locally and it should install all the necessary ingredients for testing.
2. Run Mocha tests: ```grunt test```

The tests are written in the files available in the  ```test/``` folder.

## Issues

Please browse our [existing issues](https://github.com/smart-on-fhir/cardiac-risk-app/issues) before logging new issues

## Contributing

See [CONTRIBUTING.md](CONTRIBUTING.md)

## License

Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License. You may obtain a copy of the License at

http://www.apache.org/licenses/LICENSE-2.0 Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions and limitations under the License.

CHANGELOG.md
# CHANGELOG

* 1.0.0
    - Implemented new UI design.
    - Updated folder structures to separate out code.
    - Added JavaScript files using the MVC pattern.
    - Updated project to use LESS to build CSS.
    - Added grunt support for build and tests.
    - Added testing using Mocha and PhantomJS.



test/test_setup.js

var sinonSandbox;
beforeEach(function () {
	sinonSandbox = sinon.sandbox.create();
});

afterEach(function () {
	sinonSandbox.restore();
});

var setPatientInfo = function(gender,age,hsCRP,totCholesterol,ldl,hdl,sysBP,smoker,familyHistory,patientInfo) {
	if (patientInfo === undefined) patientInfo = {};
	patientInfo.firstName = 'John';
	patientInfo.lastName = 'Doe';
	patientInfo.gender = gender;
	patientInfo.age = age;

	patientInfo.hsCReactiveProtein = hsCRP;
	patientInfo.totalCholesterol = totCholesterol;
	patientInfo.hdl = hdl;
	patientInfo.ldl = ldl;

	patientInfo.systolicBloodPressure = sysBP;
	var relatedFactors = {};
	relatedFactors.smoker = smoker;
	relatedFactors.familyHeartAttackHistory = familyHistory;
	patientInfo.relatedFactors = relatedFactors;
	return patientInfo;
};


test/static_range_slider_test.js
describe('StaticRangeSlider', function() {
	describe('buildRangeSliderDataModel', function(){
		it('returns a data model built for range slider when gender is male', function(){
			CardiacRisk.patientInfo = setPatientInfo('male',59,0.5,160,100,60,119,false,false);
			buildRangeSliderDataModel();

			expect(CardiacRisk.graphData).to.be.an('object');

			expect(CardiacRisk.graphData.crpSliderData).to.be.an('object');
			expect(CardiacRisk.graphData.crpSliderData.id).to.be.equal('cReactiveProteinSlider');
			expect(CardiacRisk.graphData.crpSliderData.titleLeft).to.be.equal('C Reactive Protein');
			expect(CardiacRisk.graphData.crpSliderData.titleRight).to.be.equal('mg/L');
			expect(CardiacRisk.graphData.crpSliderData.lowerBound).to.be.equal(0.03);
			expect(CardiacRisk.graphData.crpSliderData.upperBound).to.be.equal(20);
			expect(CardiacRisk.graphData.crpSliderData.barBoundsLowDisplay).to.be.equal('Low');
			expect(CardiacRisk.graphData.crpSliderData.barBoundsHighDisplay).to.be.equal('High');
			expect(CardiacRisk.graphData.crpSliderData.toolTipData).to.be.an('object');
			expect(CardiacRisk.graphData.crpSliderData.toolTipData.keys).to.eql(['Low', 'Moderate', 'High']);
			expect(CardiacRisk.graphData.crpSliderData.toolTipData.values).to.eql(['0.03 - 0.9', '1 - 2.9', '3 - 20']);
			expect(CardiacRisk.graphData.crpSliderData.toolTipData.styleClass).to.equal('tooltipsterCardiacRiskCRP');

			expect(CardiacRisk.graphData.totalCholesterolSliderData).to.be.an('object');
			expect(CardiacRisk.graphData.totalCholesterolSliderData.id).to.be.equal('totalCholesterolSlider');
			expect(CardiacRisk.graphData.totalCholesterolSliderData.titleLeft).to.be.equal('Total Cholesterol');
			expect(CardiacRisk.graphData.totalCholesterolSliderData.titleRight).to.be.equal('mg/dL');
			expect(CardiacRisk.graphData.totalCholesterolSliderData.lowerBound).to.be.equal(140);
			expect(CardiacRisk.graphData.totalCholesterolSliderData.upperBound).to.be.equal(401);
			expect(CardiacRisk.graphData.totalCholesterolSliderData.barBoundsLowDisplay).to.be.equal('Desirable');
			expect(CardiacRisk.graphData.totalCholesterolSliderData.barBoundsHighDisplay).to.be.equal('High');
			expect(CardiacRisk.graphData.totalCholesterolSliderData.toolTipData).to.be.an('object');
			expect(CardiacRisk.graphData.totalCholesterolSliderData.toolTipData.keys).to.eql(['Desirable', 'Borderline High', 'High']);
			expect(CardiacRisk.graphData.totalCholesterolSliderData.toolTipData.values).to.eql(['140 - 199', '200 - 239', '240 - 401']);
			expect(CardiacRisk.graphData.totalCholesterolSliderData.toolTipData.styleClass).to.equal('tooltipsterCardiacRiskTotalCholesterol');


			expect(CardiacRisk.graphData.ldlSliderData).to.be.an('object');
			expect(CardiacRisk.graphData.ldlSliderData.id).to.be.equal('ldlBadCholesterolSlider');
			expect(CardiacRisk.graphData.ldlSliderData.titleLeft).to.be.equal('LDL "Bad" Cholesterol');
			expect(CardiacRisk.graphData.ldlSliderData.titleRight).to.be.equal('mg/dL');
			expect(CardiacRisk.graphData.ldlSliderData.lowerBound).to.be.equal(50);
			expect(CardiacRisk.graphData.ldlSliderData.upperBound).to.be.equal(210);
			expect(CardiacRisk.graphData.ldlSliderData.barBoundsLowDisplay).to.be.equal('Optimal');
			expect(CardiacRisk.graphData.ldlSliderData.barBoundsHighDisplay).to.be.equal('High');
			expect(CardiacRisk.graphData.ldlSliderData.toolTipData).to.be.an('object');
			expect(CardiacRisk.graphData.ldlSliderData.toolTipData.keys).to.eql(['Optimal', 'Near/Above Optimal', 'Borderline High', 'High', 'Very High']);
			expect(CardiacRisk.graphData.ldlSliderData.toolTipData.values).to.eql(['50 - 100', '101 - 129', '130 - 159', '160 - 189', '190 - 210']);
			expect(CardiacRisk.graphData.ldlSliderData.toolTipData.styleClass).to.equal('tooltipsterCardiacRiskLDL');


			expect(CardiacRisk.graphData.hdlSliderData).to.be.an('object');
			expect(CardiacRisk.graphData.hdlSliderData.id).to.be.equal('hdlGoodCholesterolSlider');
			expect(CardiacRisk.graphData.hdlSliderData.titleLeft).to.be.equal('HDL "Good" Cholesterol');
			expect(CardiacRisk.graphData.hdlSliderData.titleRight).to.be.equal('mg/dL');
			expect(CardiacRisk.graphData.hdlSliderData.lowerBound).to.be.equal(30);
			expect(CardiacRisk.graphData.hdlSliderData.upperBound).to.be.equal(150);
			expect(CardiacRisk.graphData.hdlSliderData.barBoundsLowDisplay).to.be.equal('Protective');
			expect(CardiacRisk.graphData.hdlSliderData.barBoundsHighDisplay).to.be.equal('High');
			expect(CardiacRisk.graphData.hdlSliderData.toolTipData).to.be.an('object');
			expect(CardiacRisk.graphData.hdlSliderData.toolTipData.keys).to.eql(['High', 'Higher the Better', 'Low']);
			expect(CardiacRisk.graphData.hdlSliderData.toolTipData.values).to.eql(['60 - 150', '40 - 59', '30 - 39']);
			expect(CardiacRisk.graphData.hdlSliderData.toolTipData.styleClass).to.equal('tooltipsterCardiacRiskHDL');
		});

		it('returns a data model built for range slider when gender is female', function(){
			CardiacRisk.patientInfo = setPatientInfo('female',59,0.5,160,100,60,119,false,false);
			buildRangeSliderDataModel();

			expect(CardiacRisk.graphData).to.be.an('object');

			expect(CardiacRisk.graphData.crpSliderData).to.be.an('object');
			expect(CardiacRisk.graphData.crpSliderData.id).to.be.equal('cReactiveProteinSlider');
			expect(CardiacRisk.graphData.crpSliderData.titleLeft).to.be.equal('C Reactive Protein');
			expect(CardiacRisk.graphData.crpSliderData.titleRight).to.be.equal('mg/L');
			expect(CardiacRisk.graphData.crpSliderData.lowerBound).to.be.equal(0.03);
			expect(CardiacRisk.graphData.crpSliderData.upperBound).to.be.equal(20);
			expect(CardiacRisk.graphData.crpSliderData.barBoundsLowDisplay).to.be.equal('Low');
			expect(CardiacRisk.graphData.crpSliderData.barBoundsHighDisplay).to.be.equal('High');
			expect(CardiacRisk.graphData.crpSliderData.toolTipData).to.be.an('object');
			expect(CardiacRisk.graphData.crpSliderData.toolTipData.keys).to.eql(['Low', 'Moderate', 'High']);
			expect(CardiacRisk.graphData.crpSliderData.toolTipData.values).to.eql(['0.03 - 0.9', '1 - 2.9', '3 - 20']);
			expect(CardiacRisk.graphData.crpSliderData.toolTipData.styleClass).to.equal('tooltipsterCardiacRiskCRP');

			expect(CardiacRisk.graphData.totalCholesterolSliderData).to.be.an('object');
			expect(CardiacRisk.graphData.totalCholesterolSliderData.id).to.be.equal('totalCholesterolSlider');
			expect(CardiacRisk.graphData.totalCholesterolSliderData.titleLeft).to.be.equal('Total Cholesterol');
			expect(CardiacRisk.graphData.totalCholesterolSliderData.titleRight).to.be.equal('mg/dL');
			expect(CardiacRisk.graphData.totalCholesterolSliderData.lowerBound).to.be.equal(140);
			expect(CardiacRisk.graphData.totalCholesterolSliderData.upperBound).to.be.equal(401);
			expect(CardiacRisk.graphData.totalCholesterolSliderData.barBoundsLowDisplay).to.be.equal('Desirable');
			expect(CardiacRisk.graphData.totalCholesterolSliderData.barBoundsHighDisplay).to.be.equal('High');
			expect(CardiacRisk.graphData.totalCholesterolSliderData.toolTipData).to.be.an('object');
			expect(CardiacRisk.graphData.totalCholesterolSliderData.toolTipData.keys).to.eql(['Desirable', 'Borderline High', 'High']);
			expect(CardiacRisk.graphData.totalCholesterolSliderData.toolTipData.values).to.eql(['140 - 199', '200 - 239', '240 - 401']);
			expect(CardiacRisk.graphData.totalCholesterolSliderData.toolTipData.styleClass).to.equal('tooltipsterCardiacRiskTotalCholesterol');


			expect(CardiacRisk.graphData.ldlSliderData).to.be.an('object');
			expect(CardiacRisk.graphData.ldlSliderData.id).to.be.equal('ldlBadCholesterolSlider');
			expect(CardiacRisk.graphData.ldlSliderData.titleLeft).to.be.equal('LDL "Bad" Cholesterol');
			expect(CardiacRisk.graphData.ldlSliderData.titleRight).to.be.equal('mg/dL');
			expect(CardiacRisk.graphData.ldlSliderData.lowerBound).to.be.equal(50);
			expect(CardiacRisk.graphData.ldlSliderData.upperBound).to.be.equal(210);
			expect(CardiacRisk.graphData.ldlSliderData.barBoundsLowDisplay).to.be.equal('Optimal');
			expect(CardiacRisk.graphData.ldlSliderData.barBoundsHighDisplay).to.be.equal('High');
			expect(CardiacRisk.graphData.ldlSliderData.toolTipData).to.be.an('object');
			expect(CardiacRisk.graphData.ldlSliderData.toolTipData.keys).to.eql(['Optimal', 'Near/Above Optimal', 'Borderline High', 'High', 'Very High']);
			expect(CardiacRisk.graphData.ldlSliderData.toolTipData.values).to.eql(['50 - 100', '101 - 129', '130 - 159', '160 - 189', '190 - 210']);
			expect(CardiacRisk.graphData.ldlSliderData.toolTipData.styleClass).to.equal('tooltipsterCardiacRiskLDL');


			expect(CardiacRisk.graphData.hdlSliderData).to.be.an('object');
			expect(CardiacRisk.graphData.hdlSliderData.id).to.be.equal('hdlGoodCholesterolSlider');
			expect(CardiacRisk.graphData.hdlSliderData.titleLeft).to.be.equal('HDL "Good" Cholesterol');
			expect(CardiacRisk.graphData.hdlSliderData.titleRight).to.be.equal('mg/dL');
			expect(CardiacRisk.graphData.hdlSliderData.lowerBound).to.be.equal(30);
			expect(CardiacRisk.graphData.hdlSliderData.upperBound).to.be.equal(150);
			expect(CardiacRisk.graphData.hdlSliderData.barBoundsLowDisplay).to.be.equal('Protective');
			expect(CardiacRisk.graphData.hdlSliderData.barBoundsHighDisplay).to.be.equal('High');
			expect(CardiacRisk.graphData.hdlSliderData.toolTipData).to.be.an('object');
			expect(CardiacRisk.graphData.hdlSliderData.toolTipData.keys).to.eql(['High', 'Higher the Better', 'Low']);
			expect(CardiacRisk.graphData.hdlSliderData.toolTipData.values).to.eql(['60 - 150', '50 - 59', '30 - 49']);
			expect(CardiacRisk.graphData.hdlSliderData.toolTipData.styleClass).to.equal('tooltipsterCardiacRiskHDL');
		});
	});
});

Gruntfile.js
module.exports = function (grunt) {

  require('matchdep').filterDev('grunt-*').forEach(grunt.loadNpmTasks);

  // Project configuration.
  grunt.initConfig({
    pkg: grunt.file.readJSON('package.json'),

    availabletasks: {
      tasks: {
        options: {
          filter: 'exclude',
          tasks: ['availabletasks', 'default', 'concat', 'uglify']
        }
      }
    },

    concat: {
      cardiac_risk: {
        files: {
          'build/js/cardiac_risk.js': [
            'src/js/cardiac_risk.js',
            'src/js/cardiac_risk_controller.js',
            'src/js/static_range_slider.js',
            'src/js/static_range_slider_controller.js'
          ]
        }
      }
    },

    mocha_phantomjs: {
      all: ['test/*.html']
    },

    uglify: {
      cardiac_risk: {
        files: {
          'build/js/cardiac_risk.min.js': ['build/js/cardiac_risk.js']
        }
      }
    },

    jshint: {
      files: [
        'Gruntfile.js' ,
        'build/js/cardiac_risk.js',
        'test/*.js'],
      options: {
        globals: {
          jQuery: true,
          mocha: true,
          phantom: true,
          force: true

        }
      }
    },

    less: {
      cardiac_risk: {
        options: {
          paths: ['src/less']
        },
        files: {
          'src/stylesheets/cardiac_risk.css': ['src/less/cardiacRisk.less', 'src/less/tooltipster-CardiacRisk.less', 'src/stylesheets/tooltipster.css']
        }
      }
    },

    cssmin: {
      cardiac_risk: {
        files: {
          'build/stylesheets/cardiac_risk.css': ['src/stylesheets/cardiac_risk.css']
        }
      }
    }

  });

  grunt.registerTask('default', ['availabletasks']);

  grunt.registerTask('test', ['mocha_phantomjs']);

  grunt.registerTask('build', ['less', 'cssmin', 'concat', 'jshint', 'uglify']);

};

src/js/static_range_slider_controller.js
/**
 * Generic method to create the range slider component.
 * @param data This is the data structure needed to build the component.
 *   Structure :
 *      id : The div id to be use to reference the component.
 *      titleLeft : Title text for the left header.
 *      titleRight : Title text for the right header.
 *      lowerBound : Lower bound value for the bar calculations.
 *      upperBound : Upper bound value for the bar calculations.
 *      barBoundsLowDisplay : Display text for the footer left bar bounds.
 *      barBoundsHighDisplay : Display text for the footer right bar bounds.
 *      toolTipData :
 *      {
            "keys" : Array of left column text.
                    Eg. ["Low", "Moderate", "High"]
            "values": Array of right column text.
                    Eg. ["0.03 - 0.9", "1 - 2.9", "3 - 20"]
            "styleClass": Styling class.
        }
 *      barHeight : Height to be used for the bar component of the graph.
 */
function generateRangeSlider(data) {
    var $sliderDiv = $('#' + data.id);
    $sliderDiv.append(
        '<header id=' + data.id + 'Header' + ' class="rangeSliderHeaderStyle">' +
            '<div>' +
                '<span id=' + data.id + 'TitleLeft' + ' class="headerLeftStyle">'+ data.titleLeft +'</span>' +
                '<span id=' + data.id + 'Image' + ' class="iconStyle fa fa-info-circle"></span>' +
        '   </div>' +
            '<span id=' + data.id + 'TitleRight' + '>'+ data.titleRight +'</span>' +
        '</header>' +
        '<div class="rangeSliderContentStyle">' +
            '<div id=' + data.id + 'Bar' + ' class="rangeSliderBarStyle"></div>' +
            '<div id=' + data.id + 'Thumb' + ' class="rangeSliderThumbStyle">'+ data.value +'</div>' +
            '<div id=' + data.id + 'ThumbDisplayText' + ' class="rangeSliderThumbDisplayTextStyle">' + data.thumbDisplayText + '</div>' +
        '</div>' +
        '<footer id=' + data.id + 'Footer' + ' class="rangeSliderFooterStyle">' +
            '<div class="footerLeftStyle">' +
                '<span id=' + data.id + 'FooterLeft' + '>'+ data.barBoundsLowDisplay +'</span>' +
            '</div>' +
            '<div class="footerRightStyle">' +
                '<span id=' + data.id + 'FooterRight' + '>'+ data.barBoundsHighDisplay +'</span>' +
            '</div>' +
        '</footer>'
    );
    updateBarHeight(data.id, data.barHeight);
    updateToolTips(data.id, data.toolTipData);
    updateThumbPosition(data.id, data.value, data.lowerBound, data.upperBound);
}

/**
 * Method to update the content of tool tip for the more information icon.
 * @param id Indicates the div id to be used to update the tooltip.
 * @param toolTipData
 */
function updateToolTips(id, toolTipData) {
    var contentString = '<table class="'+ toolTipData.styleClass + '">';
    var tableRows = '';

    for (step = 0; step < toolTipData.keys.length ;step++) {
        tableRows = tableRows.concat('<tr>');
        tableRows = tableRows.concat(
            '<td class="zeroMargin">'+ toolTipData.keys[step] +'</td>' +
            '<td> </td>' +
            '<td class="zeroMargin">' + toolTipData.values[step] + '</td>'
        );
        tableRows = tableRows.concat('</tr>');
    }
    contentString = contentString.concat(tableRows, '</table>');

    $('#' + id + 'Image').tooltipster({
        contentAsHTML : true,
        delay : 100,
        theme: 'tooltipster-CardiacRisk',
        content: $(contentString)
    });
}

/**
 * Method to update the bar height in the graph.
 * @param id : Indicates the div id to be used to update the tooltip.
 * @param barHeight : Indicates the height to use.
 */
function updateBarHeight(id, barHeight) {
    $('#' + id + 'Bar').css('height', '');
    $('#' + id + 'Bar').css('height', barHeight);
    if (barHeight < 12)
    {
        $('#' + id + 'Header').css('margin-bottom',2 * barHeight);
        $('#' + id + 'Thumb').addClass('barHeightThumbAdjustment');
        $('#' + id + 'ThumbDisplayText').addClass('barHeightThumbTextAdjustment');
        $('#' + id + 'Footer').addClass('barFooterAdjustment');
    }
}

/**
 * Method to change background color of the bar.
 * @param id : Indicates the div id to be used to update the tooltip.
 * @param color
 */
function changeBarBackgroundColor(id, color) {
    $('#' + id + 'Bar').css('background-color', '');
    $('#' + id + 'Bar').addClass(color);
}

/**
 * Method to change the thumb's background color
 * @param id : Indicates the div id to be used to update the tooltip.
 * @param color
 */
function changeThumbBackgroundColor(id, color) {
    $('#' + id + 'Thumb').css('background-color', '');
    $('#' + id + 'Thumb').addClass(color);
}

/**
 * Method to change the thumb's text color
 * @param id : Indicates the div id to be used to update the tooltip.
 * @param color
 */
function changeThumbTextColor(id, color) {
    if (color.length) {
        $('#' + id + 'Thumb').removeClass('rangeSliderThumbStyle').addClass(color);
    }
}

/**
 * Method to update the thumb's left position based on the data value.
 * @param id : Indicates the div id to be used to update the tooltip.
 * @param value : The value indicating the actual data value.
 * @param lowerBound : Value for the lower bound.
 * @param upperBound : Value for the upper bound.
 */
function updateThumbPosition(id, value, lowerBound, upperBound) {
    var thumbOffset = 0;
    if (value <= lowerBound) {
        thumbOffset = 0;
        changeThumbPositionToPosition(id, thumbOffset);
        changeThumbValueTextToPosition(id, thumbOffset);
        return;
    }

    var $barWidth = $('#' + id + 'Bar').width();
    var $thumb = $('#' + id + 'Thumb');
    var thumbWidthAdjustment = $thumb.width() +
        parseFloat($thumb.css('padding-left').replace(/[^-\d.]/g, '')) +
        parseFloat($thumb.css('padding-right').replace(/[^-\d.]/g, '')) +
        parseFloat($thumb.css('borderLeftWidth').replace(/[^-\d.]/g, '')) +
        parseFloat($thumb.css('borderRightWidth').replace(/[^-\d.]/g, ''));
    var valueToConsider = value;
    if (value > upperBound) {
        valueToConsider = upperBound;
    }
    var position = ((valueToConsider - lowerBound)/(upperBound - lowerBound)) * $barWidth;
    thumbOffset = position - thumbWidthAdjustment;
    if (thumbOffset < 0) {
        thumbOffset = 0;
    }
    changeThumbPositionToPosition(id, thumbOffset);
    changeThumbValueTextToPosition(id, thumbOffset);
}

/**
 * Method to change the thumb's value text position. This is calculated based on the thumb's position.
 * If this calculated position overlaps the left bounds text or right bounds text then the left or right bounds
 * labels are hidden.
 * @param id : Indicates the div id to be used to update the tooltip.
 * @param offset : Indicates the thumb's left offset.
 */
function changeThumbValueTextToPosition(id, offset) {

    var $thumb = $('#' + id + 'Thumb');
    var $barWidth = $('#' + id + 'Bar').width();
    var $thumbDisplayWidth = $('#' + id + 'ThumbDisplayText').width();
    var thumbWidth = $thumb.width() +
        parseFloat($thumb.css('padding-left').replace(/[^-\d.]/g, '')) +
        parseFloat($thumb.css('padding-right').replace(/[^-\d.]/g, '')) +
        parseFloat($thumb.css('borderLeftWidth').replace(/[^-\d.]/g, '')) +
        parseFloat($thumb.css('borderRightWidth').replace(/[^-\d.]/g, ''));
    var offsetForDisplay = offset + (thumbWidth / 2) - ($thumbDisplayWidth / 2);
    if (offsetForDisplay < 0) offsetForDisplay = 0;
    if (offsetForDisplay > ($barWidth - $thumbDisplayWidth)) offsetForDisplay = $barWidth - $thumbDisplayWidth;

    $('#' + id + 'ThumbDisplayText').css('left', '');
    $('#' + id + 'ThumbDisplayText').css('left', offsetForDisplay);

    if (offsetForDisplay >= 0 && offsetForDisplay <= $('#' + id + 'FooterLeft').width()) {
        $('#' + id + 'FooterLeft').css('visibility', 'hidden');
    }
    else {
        if ((offsetForDisplay + $thumbDisplayWidth) <= $barWidth && (offsetForDisplay + $thumbDisplayWidth) >= $barWidth - $('#' + id + 'FooterRight').width()) {
            $('#' + id + 'FooterRight').css('visibility', 'hidden');
        }
    }
}

/**
 * Method to change the thumb's position.
 * @param id : Indicates the div id to be used to update the tooltip.
 * @param offset : Indicates the value to position the thumb.
 */
function changeThumbPositionToPosition(id, offset) {
    $('#' + id + 'Thumb').css('left', '');
    $('#' + id + 'Thumb').css('left', offset);
}


src/js/static_range_slider.js

/**
 * This method builds all the data required to display graphs for lab values.
 * Any updates to this model will reflect on the graphs being drawn on the UI.
 */
function buildRangeSliderDataModel() {
	var graphData = {};

	var crpSliderData = {};
	crpSliderData.id = 'cReactiveProteinSlider';
	crpSliderData.titleLeft = 'C Reactive Protein';
	crpSliderData.titleRight = 'mg/L';
	crpSliderData.lowerBound = 0.03;
	crpSliderData.upperBound = 20;
	crpSliderData.barBoundsLowDisplay = 'Low';
	crpSliderData.barBoundsHighDisplay = 'High';
	crpSliderData.toolTipData = {
		'keys' : ['Low', 'Moderate', 'High'],
		'values': ['0.03 - 0.9', '1 - 2.9', '3 - 20'],
		'styleClass': 'tooltipsterCardiacRiskCRP'
	};
	graphData.crpSliderData = crpSliderData;

	var totalCholesterolSliderData = {};
	totalCholesterolSliderData.id = 'totalCholesterolSlider';
	totalCholesterolSliderData.titleLeft = 'Total Cholesterol';
	totalCholesterolSliderData.titleRight = 'mg/dL';
	totalCholesterolSliderData.lowerBound = 140;
	totalCholesterolSliderData.upperBound= 401;
	totalCholesterolSliderData.barBoundsLowDisplay = 'Desirable';
	totalCholesterolSliderData.barBoundsHighDisplay = 'High';
	totalCholesterolSliderData.toolTipData = {
		'keys' : ['Desirable', 'Borderline High', 'High'],
		'values': ['140 - 199', '200 - 239', '240 - 401'],
		'styleClass': 'tooltipsterCardiacRiskTotalCholesterol'
	};
	graphData.totalCholesterolSliderData = totalCholesterolSliderData;

	if (CardiacRisk.patientInfo.ldl !== undefined) {
		var ldlSliderData = {};
		ldlSliderData.id = 'ldlBadCholesterolSlider';
		ldlSliderData.titleLeft = 'LDL "Bad" Cholesterol';
		ldlSliderData.titleRight = 'mg/dL';
		ldlSliderData.lowerBound = 50;
		ldlSliderData.upperBound = 210;
		ldlSliderData.barBoundsLowDisplay = 'Optimal';
		ldlSliderData.barBoundsHighDisplay = 'High';
		ldlSliderData.toolTipData = {
			'keys' : ['Optimal', 'Near/Above Optimal', 'Borderline High', 'High', 'Very High'],
			'values': ['50 - 100', '101 - 129', '130 - 159', '160 - 189', '190 - 210'],
			'styleClass': 'tooltipsterCardiacRiskLDL'
		};
		ldlSliderData.barHeight = 6;
		graphData.ldlSliderData = ldlSliderData;
	}

	var hdlSliderData = {};
	hdlSliderData.id = 'hdlGoodCholesterolSlider';
	hdlSliderData.titleLeft = 'HDL "Good" Cholesterol';
	hdlSliderData.titleRight = 'mg/dL';
	hdlSliderData.lowerBound = 30;
	hdlSliderData.upperBound = 150;
	hdlSliderData.barBoundsLowDisplay = 'Protective';
	hdlSliderData.barBoundsHighDisplay = 'High';
	if (CardiacRisk.patientInfo.gender === 'male') {
		hdlSliderData.toolTipData = {
			'keys' : ['High', 'Higher the Better', 'Low'],
			'values': ['60 - 150', '40 - 59', '30 - 39'],
			'styleClass': 'tooltipsterCardiacRiskHDL'
		};
	}
	else if (CardiacRisk.patientInfo.gender === 'female') {
		hdlSliderData.toolTipData = {
			'keys' : ['High', 'Higher the Better', 'Low'],
			'values': ['60 - 150', '50 - 59', '30 - 49'],
			'styleClass': 'tooltipsterCardiacRiskHDL'
		};
	}
	hdlSliderData.barHeight = 6;

	graphData.hdlSliderData = hdlSliderData;
	CardiacRisk.graphData = graphData;
}